Connectors/Rhino/Speckle.Connectors.GrasshopperShared/Parameters/SpeckleBlockInstanceWrapper.ModelObjects.cs

```diff
@@ -0,0 +1,219 @@
+#if RHINO8_OR_GREATER
+using Grasshopper.Kernel.Types;
+using Grasshopper.Rhinoceros.Model;
+using Rhino;
+using Rhino.DocObjects;
+using Rhino.Geometry;
+using Speckle.Connectors.GrasshopperShared.HostApp;
+using Speckle.Sdk.Common;
+using Speckle.Sdk.Models.Instances;
+
+namespace Speckle.Connectors.GrasshopperShared.Parameters;
+
+public partial class SpeckleBlockInstanceWrapperGoo
+{
+  private bool CastFromModelObject(object source)
+  {
+    switch (source)
+    {
+      case InstanceReferenceGeometry instanceRef:
+        return CreateFromInstanceReference(instanceRef);
+
+      case GH_InstanceReference ghInstanceRef:
+        return ghInstanceRef.Value != null && CreateFromInstanceReference(ghInstanceRef.Value);
+
+      // When implementing nested blocks support, discovered that nested blocks coming from Rhino arrive as ModelObjects
+      // containing InstanceReferenceGeometry.
+      case ModelObject modelObject:
+        return CreateFromModelObject(modelObject);
+
+      default:
+        return false;
+    }
+  }
+
+  private bool CastToModelObject<T>(ref T target)
+  {
+    switch (target)
+    {
+      case GH_InstanceReference:
+        if (Value == null)
+        {
+          return false;
+        }
+
+        if (Value.Definition == null)
+        {
+          var minimalInstanceRef = new InstanceReferenceGeometry(Guid.Empty, Value.Transform);
+          target = (T)(object)new GH_InstanceReference(minimalInstanceRef);
+          return true;
+        }
+
+        var modelInstanceDef = CreateModelInstanceDefinition(Value.Definition);
+        if (modelInstanceDef == null)
+        {
+          return false;
+        }
+
+        var instanceRefGeo = new InstanceReferenceGeometry(Guid.Empty, Value.Transform);
```

Comment:
This guid.empty should be checked, what happens when you:
- have a Model Instance in GH
- cast it to Speckle Instance
- cast it back to model instance

Seems like this mutates the def id.

Reply:
absolutely right - this was causing definition ID mutation during round-trip casting. [fix: preserve id in SpeckleBlockInstanceWrapper round-trip casting](https://github.com/specklesystems/speckle-sharp-connectors/pull/949/commits/4d81bce0eada4261b60e1e1005925d8a57dd6d84)